Conditional random fields

bramvandeneijnden opened the pull request:
extension of random field generator with random field conditioning.

Uses `sklear` as additional requirement for efficient kriging, which is not available efficiently enough in gstools.

bramvandeneijnden committed:
added first version of conditioned random field generator

new file:   random_fields/generate_conditioned_field.py

bramvandeneijnden committed:
corrections, clean-up, efficiency and commenting

modified:   random_fields/generate_conditioned_field.py

bramvandeneijnden committed:
Added tutorial:
 Changes to be committed:
	modified:   ../README.md
	new file:   Makefile
	renamed:    static/random_field.png -> _static/random_field.png
	new file:   _static/random_field_1.png
	new file:   _static/random_field_2.png
	new file:   _static/random_field_3.png
	renamed:    static/random_field_3D.png -> _static/random_field_3D.png
	new file:   authors.rst
	new file:   conf.py
	new file:   contributions.rst
	new file:   index.rst
	new file:   installation.rst
	new file:   make.bat
	new file:   requirements.txt
	new file:   theory.rst
	new file:   tutorials.rst
	modified:   ../random_fields/generate_conditioned_field.py
	modified:   ../random_fields/generate_field.py
	modified:   ../random_fields/utils.py

bramvandeneijnden committed:
Extension of tutorial to 3D conditioned random fields

Changes committed:
	new file:   _static/conditioned_random_field_3D.png
	new file:   _static/kriging_mean_2D.png
	new file:   _static/kriging_mean_3D.png
	new file:   _static/kriging_std_2D.png
	new file:   _static/kriging_std_3D.png
	new file:   _static/random_field_4.png
	new file:   _static/random_field_5.png
	modified:   index.rst
	modified:   tutorials.rst
	modified:   ../random_fields/utils.py

bramvandeneijnden committed:
added sklearn to the requirements

Changes to be committed:
	modified:   tutorials.rst
	modified:   ../requirements.txt

bramvandeneijnden committed:
deleted:    random_fields/generate_conditioned_field.py

bramvandeneijnden committed:
modified:   setup.cfg

bramvandeneijnden committed:
modified:   setup.cfg  (sklearn --> scikit-learn)

aronnoordam commented on docs/contributions.rst:
```diff
@@ -0,0 +1,99 @@
+            model.project_parameters,
+            Path(mesh_file_name).stem,
+            materials_file_name,
+            model.get_all_model_parts(),
```
this function and the class above do not exists within the RF repos, can you replace it with an existing function?

bramvandeneijnden replied:
updated with existing function

aronnoordam commented on docs/installation.rst:
```diff
@@ -0,0 +1,10 @@
+-------------------
+
+
+To be copied from STEM
```
```suggestion
Work in progress
```

aronnoordam commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
 numpy==1.24.2
 gstools==1.5.0
 matplotlib>=3.7.0
+sklearn
```
missing version, also sklearn is deprecated, i think it should be scikit-learn

aronnoordam replied:
also the dependency should be added to requirements_dev.txt

bramvandeneijnden replied:
scikit-learn corrected ==1.5.1

I tried to set matplotlib version to the one used in the STEM package (3.7.4), but this breaks one of the tests on the plotting utils...

aronnoordam commented on setup.cfg:
```diff
@@ -17,6 +17,7 @@ install_requires =
     numpy>=1.24
     gstools>=1.5.0
     matplotlib>=3.7.1
+    scikit-learn
```
missing version

bramvandeneijnden replied:
fixed

aronnoordam commented on docs/installation.rst:
```diff
@@ -0,0 +1,10 @@
+RandomFields installation
+=================
+
+.. _python_stem:
```
```suggestion
.. _python_random_fields:
```

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -4,6 +4,12 @@
 import gstools as gs
```
```suggestion
import numpy.typing as npt
import gstools as gs
```

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+
+    def set_conditioning_points(self,
+                                points: np.ndarray,
+                                values: np.ndarray,
```
```suggestion
                                points: npt.NDArray[np.float64],
                                values: npt.NDArray[np.float64],
```

aronnoordam replied:
furthermore, can you maintain the style, as described in the contributions section in the docs? I know the old code is not like that, but for new code, i think its better if its done now

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        self.noise_level = noise_level
+
+        if max(points.shape) > 2000:
+            print('too many conditioning points!')
```
why is this? And can you described it in a comment

aronnoordam replied:
```suggestion
            raise Exception(f'Too many conditioning points! There are {max(points.shape)} points, while the maximum allowed amount is {max_points}')
```

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        """
+        self.noise_level = noise_level
+
+        if max(points.shape) > 2000:
```
either check for points.shape[0] or points.shape[1].  because i think they are always n by ndim or ndim by n

aronnoordam replied:
```suggestion
        max_points = 2000
        if max(points.shape) > max_points:
```

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -4,6 +4,12 @@
+from sklearn.gaussian_process import GaussianProcessRegressor as GPR
+from sklearn.gaussian_process.kernels import RBF, \
+        ConstantKernel as C, \
+        WhiteKernel as W, Matern
```
can you just keep the names ConstantKernel and WhiteKernel, else later in the code its hard to read what is going on

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        # correct the length scales between libraries
+        if self.random_field_model_name == 'Gaussian':
+            ls_sklearn  = scale_fluctuation * 2/np.pi
+            self.KrigingKernel = W(noise_level) + RBF(length_scale = ls_sklearn)    
```
variables should be written in snake_case. Also KrigingKernel is only used in this function, so i dont think the "self" is needed. so: kriging_kernel

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+            ls_sklearn  = scale_fluctuation / np.sqrt(1/2)
+            self.KrigingKernel = W(noise_level) + Matern(length_scale = ls_sklearn)    
+
+        self.GaussianProcess = GPR(kernel = self.KrigingKernel)
```
gaussian_process

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+
+        self.GaussianProcess = GPR(kernel = self.KrigingKernel)
+        self.GaussianProcess.optimizer = None
+        self.GaussianProcess.fit(points,(values - self.mean) / np.sqrt(self.variance)  )
```
why do you normalise?

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        self.kriging_std = None
+
+
+    def generate_conditioned(self, nodes: np.ndarray) -> None:
```
```suggestion
    def generate_conditioned(self, nodes: npt.NDArray[np.float64]) -> None:
```

bramvandeneijnden replied:
corrected, but hot to properly document arrays in the docstrings? see my attempt in the docstrings

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        self.conditioning_values = values
+        
+        self.kriging_mean = None
+        self.kriging_std = None
```
You create multiple "self.{variable_name}"s here. These variables should be inititalised under, the __init__ function , such that all the member variables are know, and can be described from the beginning

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        self.kriging_mean = self.mean + np.sqrt(self.variance) * z_kriged_field
+        #
+        # create single random field at nodes and conditioning points
+        nodes_cpoints = np.vstack([nodes,self.conditioning_points])
```
can you write out the variable names a bit more, e.g. nodes_cond_points. Also for example z_cond_rf_nodes instead of z_crf_nodes. This makes the code a bit more understandable

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        nodes_cpoints = np.vstack([nodes,self.conditioning_points])
+        self.generate(nodes_cpoints)
+
+        z_crf_nodes = (self.random_field.field[:nodes.shape[0]] - self.mean) / np.sqrt(self.variance) 
```
can you describe what you are doing

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
+        
+        #
+        # create kriged mean field of random field 
+        GPrf = GPR(kernel = self.GaussianProcess.kernel_)
```
snake_case

aronnoordam reviewed (commented):
few comments about style and some stuff what is unclear for me.

Furthermore can you add tests for the added functionalities?

bramvandeneijnden committed:
Update docs/installation.rst

Co-authored-by: aronnoordam <51492202+aronnoordam@users.noreply.github.com>

bramvandeneijnden committed:
Update random_fields/generate_field.py

Co-authored-by: aronnoordam <51492202+aronnoordam@users.noreply.github.com>

bramvandeneijnden committed:
Changes for merge `Conditional random fields`

	modified:   docs/contributions.rst
	modified:   docs/installation.rst
	modified:   random_fields/generate_field.py
	modified:   random_fields/utils.py
	modified:   requirements_dev.txt
	new file:   tests/data/kriging_mean_3D.txt
	new file:   tests/data/kriging_std_3D.txt
	modified:   tests/test_random_field.py

aronnoordam commented on random_fields/generate_field.py:
```diff
@@ -91,6 +145,128 @@ def generate(self, nodes: np.ndarray) -> None:
+        
+        #
+        # create kriged mean field of random field 
+        GP_RF = GPR(kernel = self.GaussianProcess.kernel_)
```
lower case

aronnoordam commented on tests/test_random_field.py:
```diff
@@ -86,14 +86,70 @@ def test_distribution_RF_unstruc():
+    np.testing.assert_array_almost_equal(max_error_mean, 0., decimal=4)
+    np.testing.assert_array_almost_equal(max_error_std, 0., decimal=4)
+
+
+test_conditioned_RF_3D_mean_variance()
```
```suggestion
```

aronnoordam commented on tests/test_random_field.py:
```diff
@@ -86,14 +86,70 @@ def test_distribution_RF_unstruc():
+    rf.conditioned_random_field
+    rf.kriging_mean
+    rf.kriging_std
+
+    ## to regenerate new reference solutions
+    # np.savetxt('./data/kriging_mean_3D.txt',rf.kriging_mean)
+    # np.savetxt('./data/kriging_std_3D.txt',rf.kriging_std)
```
```suggestion
```

bramvandeneijnden committed:
changes accounting for review comments:
- style changes
- update docstrings
- extended test

        modified:   docs/contributions.rst
	modified:   random_fields/generate_field.py
	modified:   random_fields/utils.py
	modified:   requirements.txt
	modified:   requirements_dev.txt
	new file:   tests/data/conditioned_rf_3D.txt
	deleted:    tests/data/kriging_mean_3D.txt
	deleted:    tests/data/kriging_std_3D.txt
	modified:   tests/test_random_field.py

bramvandeneijnden committed:
modified:   requirements.txt
	modified:   requirements_dev.txt

bramvandeneijnden commented on random_fields/generate_field.py:
```diff
@@ -28,26 +28,25 @@ class RandomFields():
+        - z_kriged_field (array): standard-normal kriged random field
+        - gaussian_process (:class:sklearn.gaussian_process.GaussianProcessRegressor)
+        - conditioning_points (array): coordinates of the conditioning points, shape (:,`self.n_dims`)
+        - conditioning_values (array): values of the conditioning points, shape (:,)
```
How to document arrays in the doc strings? I've seen "Sequence(float)" here and there... 

and what about shapes and defaults?
